keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,6 +1069,161 @@ the remaining requests in that queue start getting faster service.  In
+of the request.  Instead of saying that every request is consuming a single
+unit of concurrency, we allow for a request to consume `<width>` units of
+concurrency while being processed.
+
+This basically means, that the cost of processing a given request is no
+longer reflected by its `<processing latency>` and instead its cost is now
+equal to `<width> x <processing latency>`.  The rationale behind it is that
+the request is now consuming `<width>` concurrency units for the duration
+of its processing.
+
+While in theory the `width` can be an arbitrary non-integer number, for
+practical reasons, we will assume it actually is an integer.  Given that
+our estimations here are very rough anyway that seems a reasonable
+simplification that makes dispatching the budget a bit simpler.
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we want to start
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = A
+  width_ram(N) = B * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  To stay on the safe
+side, we can just protect every resource independently, which translates to
+to the following function:
+```
+  width(n) = max(A, B * N)
```

Comment:
This is one of the things I considered, but it's not a valid way of collapsing a multi dimensional space into a single dimension. A and B are not in the same units, even if you compensate by e.g. putting them in percent of available, it doesn't solve the problem.

Suppose you have 100 CPU and 100 RAM, {100, 100}. Currently the requests {60, 10} and {10, 40} are in flight. By the given formula, the width is 100 and nothing else fits, however there's actually {30, 50} free space! But if you permit over-subscribing on width, then requests that aren't so lopsided (e.g. {30, 30}) will cause the incompressible resource (RAM) to blow up.

Reply:
BTW my next step was to figure that maybe taking the vector magnitude would be a slightly better one-dimensional summary, and I still think that's probably right. E.g. `sqrt(A^2, (B*N)^2)`-- however this more or less ends up being linear in B when B dominates, hence yesterday's comment.

Reply:
> I think your main observation that I hadn't incorporated is that requests use a fair amount more CPU than RAM (proportionally).

Yeah - I think this is exactly one of the most important point (it's not emphasized enough currently).
My back-of-envelope calculations say that it will probably be for N above 10^5.

So there will be very very few requests for which the RAM part would dominate.

I agree that in a generic case we're risking causing significant underutilization (i.e. if the proportions between the two dimensions can significantly different). But I think in our case, it's pretty much always the CPU dimensions that will be dominating.
Hence - I think it's actually ok to very occasionally have some small underutilization, as the most important thing I want to guarantee is to avoid being overloaded.


> Using the magnitude, there's not a discontinuity, it just grows very slowly with number of items until the threshold is passed.

Can you clarify this one? I think I'm not following.

> I think under either formula, width should not be integer-- the gradations we'll capture are probably quite small, especially near the threshold where RAM becomes heavier.

We will never be super accurate with our estimations so I'm not sure it actually matters. And it makes dispatching a bit easier (well maybe not much, but it's a bit more intuitive). I'm not going to fight until death for that, but I don't think it actually will change anything visibly.

> I think it's important to be clear that width is compared with concurrency, not shares (I confused myself on this for a bit).

Yes - it's about concurrency. I thought it's clear, but if not I will emphasize it more at the beginning of the text.

Reply:
> Oh, really? That's huge, we would probably get way more mileage out of requiring pagination, capping it at like 5000 items per page, and not doing this at all?

We don't support pagination for serving from watchcache, and e.g. "listing pods from your node" has an index in watchcache and will be super inefficient when listed from etcd.
But for all requests that are hitting etcd - I agree that we should probably require pagination.


> I guess I was thinking that widths of 1.5 will (should?) be fairly frequent and if you get a few of those at the same time, there's a big difference in concurrency if you round.

What I was thinking is that we will leave the width of all single-item requests as 1 (as it is now), and A I expect to be approximately ~10 for larger lists.
The question is about those small-ish lists (e.g. say there are 20 or 50 objects). For these setting width=10 might be too much and we want to use smaller values. And I agree that there is a difference between 1.5 and 2, but given that even the inflight limits are some estimations, I don't think that being a bit more accurate here will actually help.
I think that overall different approximations we will be using in those params (A and B will also be some approximations) will somehow amortize and we won't get significant gain from using non-integer numbers.

Reply:
By your estimate width=10 is assuming like 100,000 items? Or is your estimate not linear?

Ideally this would work for smaller clusters too, meaning we'd get more predictable/steady resource usage.

I'm actually quite skeptical that 100,000 items is only 10x as expensive as 1 item.

Maybe instead of deciding here, we should have a function that's like `WidthEstimatorGoesHere()`, define measurement criteria, and let people optimize the estimation. I'm fine with the first draft of this being unpaginated gets 10, other requests get ceil(# items / 10000).

Reply:
> By your estimate width=10 is assuming like 100,000 items? Or is your estimate not linear?

Saying width=10, I was assuming items>N (N=1000-ish?)
I agree that for smaller lists it should be smaller.
The main point I'm making here is that there is a cap here, because we don't parallelize individual requests.

So i actually claim that 10,000 items and 100,000 items has the same width, and the cost difference (for CPU itself) generally goes to request latency.

> Maybe instead of deciding here, we should have a function that's like WidthEstimatorGoesHere(), define measurement criteria, and let people optimize the estimation. I'm fine with the first draft of this being unpaginated gets 10, other requests get ceil(# items / 10000).

So I was trying to make a point that changing this later is easy as it is well separated. I just wanted to give some reasonable-ish starting point.
One problem with paginated lists is that they may have selectors. And for very selective ones, we may need to process many more objects to return <page size> of them. Something to keep in mind.

But yes - having a function of  ceil(#items/A) {I think A will be much smaller than 10k) capped at 10 is what sounds good enough as starting point for me.

Does that make sense to you?

Reply:
> By your estimate width=10 is assuming like 100,000 items? Or is your estimate not linear?

Saying width=10, I was assuming items>N (N=1000-ish?)
I agree that for smaller lists it should be smaller.
The main point I'm making here is that there is a cap here, because we don't parallelize individual requests.

So i actually claim that 10,000 items and 100,000 items has the same width, and the cost difference (for CPU itself) generally goes to request latency.

> Maybe instead of deciding here, we should have a function that's like WidthEstimatorGoesHere(), define measurement criteria, and let people optimize the estimation. I'm fine with the first draft of this being unpaginated gets 10, other requests get ceil(# items / 10000).

So I was trying to make a point that changing this later is easy as it is well separated. I just wanted to give some reasonable-ish starting point.
One problem with paginated lists is that they may have selectors. And for very selective ones, we may need to process many more objects to return <page size> of them. Something to keep in mind.

But yes - having a function of  ceil(#items/A) {I think A will be much smaller than 10k) capped at 10 is what sounds good enough as starting point for me.

Does that make sense to you?

Reply:
I updated the doc to reflect the above discussion. PTAL